### ISSUE TITLE
Genie seems to be a strong baseline even with very small model size

### ISSUE THREAD

**Issue by StarCycle** [1x-technologies/1xgpt#7]
Hello,

It seems that GENIE is a strong baseline even with very small model size. In the following table, results of GENIE_210M are measured again.

User | Teacher-Forced CE Loss | Teacher-Forced Token Accuracy | Autoregressive CE Loss | Autoregressive Token Accuracy | Autoregressive LPIPS | Generation Time* (secs/frame)
-- | -- | -- | -- | -- | -- | --
[1x-technologies/GENIE_210M](https://huggingface.co/1x-technologies/GENIE_210M) (--single_pass) | N/A | N/A | 3.18 | 0.318 | 0.198 | 0.076
[1x-technologies/GENIE_210M](https://huggingface.co/1x-technologies/GENIE_210M) (--maskgit_steps 8) | N/A | N/A | 3.592 | 0.257 | 0.172 | 0.698
[StarCycle/GENIE_2.6M](https://huggingface.co/StarCycle/GENIE_2.6M) (--single_pass | N/A | N/A | 3.471 | 0.288 | 0.255 | 0.003
[StarCycle/GENIE_2.6M](https://huggingface.co/StarCycle/GENIE_2.6M) (--maskgit_steps 8) | N/A | N/A | 3.806 | 0.220 | 0.217 | 0.027
[1x-technologies/Llama_1B_v0](https://huggingface.co/1x-technologies/Llama_1B_v0) | 2.46 | 0.396 | 5.08 | 0.267 | 0.23 | 2.31

Details of StarCycle/GENIE_2.6M: `--num_layers 6 --d_model 128 --num_heads 4`. Other hyperparameters are the same as GENIE_210M. The exact parameter number of GENIE_2.6M is 2655465, while GENIE_210M has 210020329. The training log is [here](https://api.wandb.ai/links/lizh9785/nznai8gt). I trained it on a single 4090 GPU for 26 hours.

Example video from GENIE_210M (a video contains [prompt frames, predicted frames, ground truth frames]):

![generated_offset0_210M](https://github.com/1x-technologies/1xgpt/assets/33491471/59eba04c-a707-4e8a-8795-26d106aebcb5)

Example video from GENIE_2.6M (a video contains [prompt frames, predicted frames, ground truth frames]):

![generated_offset0_small](https://github.com/1x-technologies/1xgpt/assets/33491471/cdf46c8d-dd6b-4527-901b-950b1d2befee)

Questions:
- The more `maskgit_steps` we have, the worse ` Autoregressive CE Loss` and `Autoregressive Token Accuracy` will be. Why?
- LLama 3 is clearly worse than GENIE. It cannot even compete with GENIE_210M with a single pass. Is my understanding correct?
- If we just look at the LPIPS metric, GENIE_2.6M with multiple maskgit steps performs better than LLama3 1B. But if we look at the video, GENIE_2.6M cannot predict what will happen after openning the door. Would you like to add more metrics apart from LPIPS?

Best,
StarCycle

**Comment by StarCycle** [comment]
@kev-zhao @ericjang

**Comment by StarCycle** [comment]
Hi @Guitaricet, I see. Thanks! So I dont need to worry about **autoregressive CE loss**

**Comment by StarCycle** [comment]
Hi @Guitaricet @kev-zhao,

Sorry to open this issue again, but I find something is still confusing to me:

- What's the meaning of "Autoregressive CE Loss" for MaskGIT? Unlike LLaMA, MaskGIT does not predict token one by one (autoregressively). It generates multiple image tokens in a single pass.
- For Genie (maskgit), the more forward steps it takes, the worse "Autoregressive CE Loss" and "Autoregressive Token Accuracy" are. Why?

Best,
StarCycle

**Comment by kev-zhao** [comment]
Hi StarCycle,

- I can see why the "autoregressive" terminology is confusing, as MaskGIT is not an autoregressive LM. Our "Autoregressive CE Loss" and "Autoregressive Token Accuracy" refer to the cross entropy loss and accuracy averaged over all 16x16 tokens of the next frame. We meant to distinguish between "teacher-forced" and "autoregressive" generation, there is an explanation of the difference [here](https://github.com/1x-technologies/1xgpt?tab=readme-ov-file#metric-details).
- One possible reason is the gap between training and inference. For example, in v0.0.1, we masked at least 50% of tokens in frames 2-16 during training, which differs from inference, where previous frames are fully unmasked and the latest frame may have less than 50% masked tokens in later inference steps. We made some changes to masking in v1.0, and quick experiments suggest that more inference steps improves token accuracy for the new baseline. I am not aware of how MaskGIT-style sampling can be used to improve likelihoods/loss, please correct me if I'm wrong. However, despite the higher accuracy, taking multiple inference steps seems to result in more copying of the previous frame, perhaps because confidences are higher for copied tokens.

**Comment by StarCycle** [comment]
Hi @kev-zhao,

I see! Thanks for the response!
